src/main/java/uk/gov/homeoffice/digital/sas/timecard/validators/timeentry/TimeEntryValidator.java

```diff
@@ -30,6 +35,87 @@ public boolean isValid(Object value, ConstraintValidatorContext context) {
         timeEntry.getActualEndTime());
 
     session.setHibernateFlushMode(FlushMode.AUTO);
-    return timeEntryClashes.isEmpty();
+    if (!timeEntryClashes.isEmpty()) {
+      ClashingProperty clashingProperty = getClashingProperty(timeEntry, timeEntryClashes);
+      var payload = timeEntryClashes.stream().map(this::transformTimeEntry)
+          .collect(Collectors.toCollection(ArrayList::new));
+
+      HibernateConstraintValidatorContext hibernateContext =
+          context.unwrap(HibernateConstraintValidatorContext.class);
+
+      hibernateContext.disableDefaultConstraintViolation();
+      hibernateContext.withDynamicPayload(payload);
+      hibernateContext
+          .buildConstraintViolationWithTemplate(
+              "Time periods must not overlap with another time period")
+          .addPropertyNode(clashingProperty.toString())
+          .addConstraintViolation();
+      return false;
+    }
+    return true;
+  }
+
+  private JSONObject transformTimeEntry(TimeEntry timeEntry) {
+    var result = new JSONObject();
+    result.put("startTime", timeEntry.getActualStartTime());
+    result.put("endTime", timeEntry.getActualEndTime());
+    result.put("timePeriodTypeId", timeEntry.getTimePeriodTypeId());
+    return result;
+  }
+
+  private ClashingProperty getClashingProperty(
+      TimeEntry timeEntry, List<TimeEntry> timeEntryClashes) {
+    var startTimeClash = false;
+    var endTimeClash = false;
+
+    for (TimeEntry timeEntryClash : timeEntryClashes) {
+      if (timeEntry.getActualStartTime().equals(timeEntryClash.getActualStartTime())) {
+        startTimeClash = true;
+      }
+
+      if (timeEntryClash.getActualEndTime() != null
+          && isStartTimeInTimeEntry(timeEntry.getActualStartTime(), timeEntryClash)) {
+        startTimeClash = true;
+      }
+
+      if (timeEntry.getActualEndTime() != null
+          && timeEntryClash.getActualEndTime() != null
+          && isEndTimeInTimeEntry(timeEntry.getActualEndTime(), timeEntryClash)) {
+        endTimeClash = true;
+      }
+
+      if (timeEntry.getActualEndTime() != null
+          && timeEntryClash.getActualEndTime() != null
+          && isStartTimeInTimeEntry(timeEntryClash.getActualStartTime(), timeEntry)
+      ) {
+        endTimeClash = true;
+      }
+    }
+
+    if (startTimeClash && endTimeClash) {
+      return ClashingProperty.startAndEndTime;
+    }
+    if (startTimeClash) {
+      return ClashingProperty.startTime;
+    }
+    if (endTimeClash) {
+      return ClashingProperty.endTime;
+    }
+
+    return null;
+  }
+
+  private boolean isEndTimeInTimeEntry(Date endTime, TimeEntry timeEntry) {
+    return (endTime.after(timeEntry.getActualStartTime())
+        && endTime.getTime() <= timeEntry.getActualEndTime().getTime());
+  }
+
+  private boolean isStartTimeInTimeEntry(Date startTime, TimeEntry timeEntry) {
```

Comment:
nit: At first glance this method name made me think it was just checking that the timeEntry had that same startTime param value. Upon inspection I can see it's checking for a clash. 
Can we slightly improve the name, same for the above method.

Reply:
Updated to `startTimeClashesWithTimeEntry` and `endTimeClashesWithTimeEntry`